Add new Blocks for structural types

haozhun opened the pull request:
Depends on ~~#2910~~ ~~#3148~~

Add ArrayBlock, InterleavedBlock, ArrayBlockBuilder, InterleavedBlockBuilder, ArrayElementBlockWriter,
LazyArrayBlock

haozhun commented:
This PR has been updated to add `clazz` parameter to getObject

cberner commented:
This needs tests

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java:
```diff
@@ -133,6 +140,18 @@
+     * operators that hold on to a range of values without holding on to the
+     * entire block.
+     */
+    Block getCompactedRegion(int position, int length);
```
We needed this early, so we merged your changes for the existing blocks and named it `copyRegion`

cberner commented:
Lets split this commit up into smaller pieces. Can you split out the changes to the interfaces and the existing blocks, from the addition of the new blocks?

cberner commented on presto-main/src/main/java/com/facebook/presto/block/BlockUtils.java:
```diff
@@ -52,7 +52,7 @@ else if (value instanceof String) {
             type.writeSlice(blockBuilder, slice, 0, slice.length());
         }
         else {
-            throw new IllegalArgumentException("Unsupported type: " + value.getClass().getName());
```
Check that `type.getJavaType()` is a reference type (it could be `int`)

haozhun replied:
If all primitive types is excluded here, what about `Integer`? I found it weird either way.

cberner commented:
Also, I think you can split the new block commit into a few pieces, one for array, one for row, and one for block list block

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractBlockListBlock.java:
```diff
@@ -0,0 +1,257 @@
+
+    public abstract BlockListBlockEncoding getEncoding();
+
+    public AbstractBlockListBlock(int columns)
```
this should be protected

cberner commented:
I think we can simplify this into just two new Blocks instead of three. Let me know if this works:

```
InterleavedBlock: a heterogeneous Block, which interleaves N blocks
Block[] blocks;
int baseOffset;
int length;

getLong(int position)
{
  int block = position % blocks.length;
  int blockPosition = position / blocks.length;
  return blocks[block].getLong(blockPosition + baseOffset);
}
```
Conceptually, this would be the native container Block for all structural types. That is, arrays would be an InterleavedBlock with a single underlying block and N positions, maps would be an InterleavedBlock with two underlying blocks and heterogeneous 2*N positions, and rows would be InterleavedBlock with N underlying blocks and N heterogeneous positions.

Then the Block that structural values would be stored in would be:
```
StructuralBlock (alternative name BlockBlock, or JaggedMatrix): is a Block of Blocks
Block[] blocks (all lengths must be equal)
int[] offsets (starting point in blocks of each position)

getObject(int position)
{
  return new InterleavedBlock(blocks, offsets[position])
}
```

haozhun commented:
I'm not sure if I understand the "two block" simplification you proposed above. Before we draw the conclusion of whether this work, I have a few questions

* It looks like you are trying to split the concept of "native container block" and that of "blocks that values would be stored in". I don't see why a block of `[1,2,3]`, `[4,5]`, and `[6]` should be different from the native container of `[[1,2,3], [4,5], [6]]`.

In addition, I don't quite see the correspondence between my implementation and the "two block" simplification you are suggesting. Can you answer the following questions for me to help me understand your simplification?
* How should the native container block of `[[1,2,3], [4,5], [6]]` look like?
* How should the native container block of `[row(1,'a'), null, row(5,'c')]` look like?
* How should the native container block of `row(1, 'a')` look like?

haozhun commented:
I don't want to prematurely split commits and later having to make major changes that spans across all commits. Therefore, I consider the following a better alternative than doing the split now:

* Split changes of existing blocks and additions of new blocks if the interface change "looks good" to you.
* Split the addition of each new abstract block if we conclude that those are the four blocks we are going to add.

cberner commented:
I was more trying to simplify things into the two concepts we need: blocks of blocks, and heterogeneous blocks. Array of array of bigint would still have the same block type as the container for array bigint - - which would be a blockblock of a blockblock of fixed width block. 

For the row example, the block containing `row(1, 'a') ` would be a blockblock of an interleaved block of a fixed width block and a variable width block. The case you mentioned above with the array would be the same, but with an extra blockblock level. Did that make sense?

haozhun commented:
It looks like your `BlockBlock` is similar to `ArrayBlock`, and your `InterleavedBlock` is similar to `BlockListBlock`.

And you are right. The container for `row(int, varchar)` is really the same as the container for `array(int)` except that the size of each element in the first case is always 2.

But I still find it necessary (for reasonable performance) to have a `ViewBlock` because `getObject` operation will be very expensive for `BlockBlock` of `InterleavedBlock`.

cberner commented:
The changes to `Block` and the existing blocks all make sense to me. Have @dain take a look at them too, but ya splitting those out sounds good.

cberner commented:
Ya, I see why you need `ViewBlock` for rows. Alternatively though, would it make sense to optimize `getRegion` by adding an offset and length to `InterleavedBlock` so that it wouldn't have to call getRegion on all its underlying blocks? If there are other uses for ViewBlock, I think it's a fine abstraction

haozhun commented:
Depend on #3148

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/block/BlockBuilder.java:
```diff
@@ -54,6 +54,20 @@
+     * Return a writer to the current entry. The caller can operate on the returned caller to incrementally build the object. This is generally more efficient than
+     * building the object elsewhere and call writeObject afterwards because a large chunk of memory could potentially be unnecessarily copied in this process.
+     */
+    default BlockBuilder nextObjectWriter()
```
Putting this function in a specific BlockBuilder and downcast when needed is not an option. All three implementations added in this pull request needs to implement this class although the only one that "really" need it is `ArrayBlock`.

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/block/BlockBuilder.java:
```diff
@@ -62,6 +62,15 @@ default BlockBuilder writeObject(Object value)
+     * Return a writer to the current entry. The caller can operate on the returned caller to incrementally build the object. This is generally more efficient than
+     * building the object elsewhere and call writeObject afterwards because a large chunk of memory could potentially be unnecessarily copied in this process.
+     */
+    default BlockBuilder nextObjectWriter()
```
Putting this function in a specific BlockBuilder and downcast when needed is not an option. All three implementations added in this pull request needs to implement this class although the only one that "really" need it is ArrayBlock.

cberner replied:
I would name this `createBlockPositionWriter`, since it can only be used to write `Object` positions that are `Block`s. @dain thoughts?

haozhun replied:
should be `element` instead of `position`?

haozhun commented:
Please review

cberner commented on presto-main/src/main/java/com/facebook/presto/block/BlockSerdeUtil.java:
```diff
@@ -0,0 +1,62 @@
+
+public class BlockSerdeUtil
+{
+    public static final MethodHandle READ_BLOCK = Reflection.methodHandle(BlockSerdeUtil.class, "readBlock", BlockEncodingSerde.class, Slice.class);
```
static import

cberner commented on presto-main/src/main/java/com/facebook/presto/block/BlockSerdeUtil.java:
```diff
@@ -0,0 +1,62 @@
+
+import java.lang.invoke.MethodHandle;
+
+public class BlockSerdeUtil
```
make utility classes `final`

cberner replied:
this class doesn't seem to be used

cberner commented on presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java:
```diff
@@ -0,0 +1,245 @@
+        assertBlock(blockBuilder.build(), expectedValues);
+    }
+
+    private static Object[] alternatingNullValues(Object[] objects)
```
there seems to be a lot of overlap with `AbstractTestBlock` I think you can remove most of it

cberner commented on presto-main/src/test/java/com/facebook/presto/block/TestInterleavedBlock.java:
```diff
@@ -0,0 +1,95 @@
+
+    private int roundUp(int position, int divisor)
+    {
+        if (roundDown(position, divisor) == position) {
```
you could simplify this `if` to `return roundDown(position + divisor - 1, divisor)`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractStructuralBlock.java:
```diff
@@ -0,0 +1,23 @@
+ */
+package com.facebook.presto.spi.block;
+
+public abstract class AbstractStructuralBlock
```
It looks like this has only one subclass, so inline it there

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/LazyArrayBlock.java:
```diff
@@ -0,0 +1,117 @@
+public class LazyArrayBlock
+        extends AbstractArrayBlock
+{
+    private Block values = null;
```
`null` is the default value

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockEncoding.java:
```diff
@@ -0,0 +1,105 @@
+        implements BlockEncoding
+{
+    public static final BlockEncodingFactory<InterleavedBlockEncoding> FACTORY = new InterleavedBlockEncodingFactory();
+    private static final String NAME = "BLOCK_LIST";
```
change this to "INTERLEAVED_BLOCK"?

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockEncoding.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    public void writeBlock(SliceOutput sliceOutput, Block block)
+    {
+        AbstractInterleavedBlock blockListBlock = (AbstractInterleavedBlock) block;
```
name this variable `interleavedBlock`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockBuilder.java:
```diff
@@ -0,0 +1,289 @@
+        extends AbstractInterleavedBlock
+        implements BlockBuilder
+{
+    private final BlockBuilderStatus blockBuilderStatus;
```
this field is unused. Is that a bug, or can it be removed?

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockBuilder.java:
```diff
@@ -0,0 +1,289 @@
+    {
+        this(
+                blockBuilderStatus,
+                types.stream().map(t -> t.createBlockBuilder(blockBuilderStatus, expectedEntries)).toArray(BlockBuilder[]::new)
```
add newline before `.map` and `.toArray`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockBuilder.java:
```diff
@@ -0,0 +1,287 @@
+
+    public InterleavedBlockBuilder(List<Type> types, BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        this(
```
the formatting looks off here. I think the argument should be on the same like as the `this(`

cberner replied:
nevermind, looks like I'm wrong

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockBuilder.java:
```diff
@@ -0,0 +1,287 @@
+    private final InterleavedBlockEncoding blockEncoding;
+
+    private int positionCount;
+    private int currentBlockId;
```
I would name this `currentBlock` or `currentBlockIndex`. `currentBlockId` sounds like it might be non-sequencial.

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockBuilder.java:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    @Override
+    protected Block getBlock(int blockId)
```
I'd just name the argument `block` or `index`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockBuilder.java:
```diff
@@ -0,0 +1,287 @@
+    @Override
+    public BlockBuilder writeByte(int value)
+    {
+        BlockBuilder blockBuilder = blockBuilders[currentBlockId];
```
put all this setup code in a helper method

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlock.java:
```diff
@@ -0,0 +1,87 @@
+        this.blocks = blocks;
+
+        BlockEncoding[] individualBlockEncodings = new BlockEncoding[columns];
+        int block0PositionCount = blocks[0].getPositionCount();
```
Add a check that `blocks` is non-empty

cberner replied:
oh nm, `super()` checks that

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,261 @@
+public abstract class AbstractInterleavedBlock
+        implements Block
+{
+    protected final int columns;
```
make this private, and have the subclasses use the getter

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlock.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Override
+    protected Block getBlock(int blockId)
```
`block` instead of `blockId`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,261 @@
+    protected AbstractInterleavedBlock(int columns)
+    {
+        if (columns == 0) {
+            throw new IllegalArgumentException("Number of blocks in BlockListBlock must not be 0.");
```
remove period at end of error message

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,261 @@
+
+    protected AbstractInterleavedBlock(int columns)
+    {
+        if (columns == 0) {
```
check that it's `> 0`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,261 @@
+     * Can only be called after the child class is initialized enough that getBlock will return the right value
+     * @return
+     */
+    protected InterleavedBlockEncoding computeBlockEncoding()
```
this is only used by the builder, so move it there

haozhun replied:
The logic was duplicated in InterleavedBlock. After removing duplication, it is no longer used only in builder.

cberner replied:
Make this a static that takes a `Block[]` as an argument, then you can remove the comment and the implicit dependency on the class being sufficiently initialized

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,261 @@
+    @Override
+    public void writePositionTo(int position, BlockBuilder blockBuilder)
+    {
+        int blockId = position % columns;
```
name all of these `block`, `index` or `blockIndex`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,261 @@
+        int positionInBlock = position / columns;
+
+        return getBlock(blockId).getLength(positionInBlock);
+
```
blank line

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,261 @@
+        int blockId = position % columns;
+        int positionInBlock = position / columns;
+
+        if (blockId < 0) {
```
add a helper for this and put it in all the other getters

cberner commented on presto-main/src/test/java/com/facebook/presto/block/TestInterleavedBlock.java:
```diff
@@ -0,0 +1,106 @@
+        return position / divisor * divisor;
+    }
+
+    private int roundUp(int position, int divisor)
```
static

cberner commented on presto-main/src/test/java/com/facebook/presto/block/TestInterleavedBlock.java:
```diff
@@ -0,0 +1,106 @@
+        assertPositionValue(block.copyRegion(roundDown(position, COLUMN_COUNT), block.getPositionCount() - roundDown(position, COLUMN_COUNT)), position % COLUMN_COUNT, expectedValue);
+    }
+
+    private int roundDown(int position, int divisor)
```
static

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,258 @@
+        return result;
+    }
+
+    protected abstract Block getValues();
```
put blank lines between these

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,258 @@
+        checkReadablePosition(position);
+
+        int positionCount = getPositionCount();
+        if (position < 0 || position >= positionCount) {
```
this check duplicates `checkReadablePosition` remove one of them

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,258 @@
+        }
+        int startValueOffset = getOffset(position);
+        int endValueOffset = getOffset(position + 1);
+        return (T) getValues().getRegion(startValueOffset, endValueOffset - startValueOffset);
```
use `clazz.cast` instead of `(T)` to get rid of the warning

haozhun replied:
I have had this question about warnings for quite some time. How do you see the warnings?

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,258 @@
+    {
+        int positionCount = getPositionCount();
+        if (position < 0 || position >= positionCount) {
+            throw new IndexOutOfBoundsException("Invalid position " + position + " in block with " + positionCount + " positions");
```
call `checkReadablePosition`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    protected abstract Block getValues();
+    protected abstract Block getOffsets();
```
Use a `Slice` instead of a `Block` since you don't need nullable positions, and a `Slice` is lighter weight

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,258 @@
+
+    protected abstract Block getValues();
+    protected abstract Block getOffsets();
+    protected abstract int getOffsetBase();
```
This offset base makes the code more complicated. I think you can remove it by:
* making the offsets Slice have length = positions + 1
* changing getRegion (non-compacting) to reuse the same values Block without calling getRegion on it
* changing getRegion (compacting) to rewrite the offsets to start at zero (you're already doing some much memory copying, that some extract computation won't hurt here)

haozhun replied:
This is a great idea.

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayElementBlock.java:
```diff
@@ -0,0 +1,211 @@
+
+import io.airlift.slice.Slice;
+
+public abstract class AbstractArrayElementBlock
```
It looks like this has only one subclass, so just inline it

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayElementBlockWriter.java:
```diff
@@ -0,0 +1,145 @@
+    @Override
+    public String toString()
+    {
+        StringBuilder sb = new StringBuilder("RowViewBlockWriter{");
```
wrong name

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayElementBlock.java:
```diff
@@ -0,0 +1,211 @@
+    @Override
+    public BlockEncoding getEncoding()
+    {
+        // RowViewBlock does not exist
```
wrong comment

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayElementBlock.java:
```diff
@@ -0,0 +1,211 @@
+
+    private Block getRegion(int position, int length, boolean compact)
+    {
+        // Although a `RowViewBlock` can only represent 1 row. Implementation of `copyRegion` is different from `getRegion` because `RowViewBlock`
```
wrong comment

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayElementBlockWriter.java:
```diff
@@ -0,0 +1,145 @@
+        throw new UnsupportedOperationException();
+    }
+
+    public void closeWriter(BlockBuilder creator)
```
Instead of making the caller of `nextObjectWriter` call this method, I would store the current writer into a member variable in `ArrayBlockBuilder.nextObjectWriter`, return `this` from `ArrayElementWriter.build()` and in `ArrayBlockBuilder.writeObject` check if the `value == currentWriter`. Then you can remove this method, and the workflow for callers will be very natural:
```
BlockBlocker array = builder.nextObjectWriter();
// write stuff to array
builder.writeObject(array.build())
```

haozhun replied:
I don't understand this one. Let's talk.

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayElementBlockWriter.java:
```diff
@@ -0,0 +1,145 @@
+        extends AbstractArrayElementBlock
+        implements BlockBuilder
+{
+    private BlockBuilder blockBuilder;
```
final

cberner replied:
you should turn on the IntelliJ warnings for this

cberner commented on presto-main/src/test/java/com/facebook/presto/block/TestArrayBlock.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    @Override
+    protected <T> void assertPositionValue(Block block, int position, T expectedValue)
```
I don't really like the idea of making this a member function, just to override it in this one class. Instead I would merge this functionality into the original code, by adding another case for `classT == Slice.class`, and probably a helper method that does all the `Slice` checks in the original code, since they're rather extensive

cberner commented on presto-main/src/test/java/com/facebook/presto/block/TestInterleavedBlock.java:
```diff
@@ -0,0 +1,106 @@
+    protected <T> void assertBlockPosition(Block block, int position, T expectedValue)
+    {
+        assertPositionValue(block, position, expectedValue);
+        assertPositionValue(block.getSingleValueBlock(position), 0, expectedValue);
```
Instead of all this code to align the regions, I think you should just all unaligned calls to `getRegion`. Conceptually, there's no reason to disallow it, and since you already have `positionCount` in `InterleavedBlock` it looks easy to allow it in the implementation

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,234 @@
+            newValues = getValues().copyRegion(startValueOffset, endValueOffset - startValueOffset);
+            int[] newOffsetsArray = new int[length + 1];
+            for (int i = 0; i <= length; i++) {
+                newOffsetsArray[i] = getOffsets().getInt((position + i) * 4) - startValueOffset;
```
simplify to `getOffset(position + i) - startValueOffset`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,234 @@
+            }
+            newOffsets = Slices.wrappedIntArray(newOffsetsArray);
+            newValueIsNull = Slices.copyOf(getValueIsNull(), position, length);
+            // In the case of compact, one could argue that is is beneficial to always set startValueOffset to 0, and adjust newOffsets accordingly.
```
Remove comment. It's out of date

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,234 @@
+    @Override
+    public int getSizeInBytes()
+    {
+        return Integer.BYTES + getValues().getSizeInBytes() + getOffsets().length() + getValueIsNull().length();
```
Why + `Integer.BYTES`? `ArrayBlock` doesn't seem to have anything that's the size of an int, and `ArrayBlockBuilder` has more

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayElementBlock.java:
```diff
@@ -0,0 +1,211 @@
+        this.start = start;
+    }
+
+    int getStart()
```
this is unused

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayElementBlock.java:
```diff
@@ -0,0 +1,211 @@
+        return getRegion(position, length, true);
+    }
+
+    private Block getRegion(int position, int length, boolean compact)
```
inline this method, since it just throws and exception. You can remove the comment too

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayElementBlock.java:
```diff
@@ -0,0 +1,211 @@
+    @Override
+    public BlockEncoding getEncoding()
+    {
+        // ArrayElementBlock does not exist
```
wrong class name in comment

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayElementBlock.java:
```diff
@@ -0,0 +1,211 @@
+    @Override
+    public int getSizeInBytes()
+    {
+        return 2 * Integer.BYTES + getBlock().getSizeInBytes();
```
It seems like this should be `getBlock().getRegion(start, getPositionCount()).getSizeInBytes()`, but that might be expensive, so I would move this to the subclass

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayElementBlock.java:
```diff
@@ -0,0 +1,211 @@
+    @Override
+    public int getRetainedSizeInBytes()
+    {
+        return 2 * Integer.BYTES + getBlock().getRetainedSizeInBytes();
```
you can remove the constant

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,256 @@
+
+    protected abstract Block getBlock(int blockIndex);
+
+    public abstract InterleavedBlockEncoding getEncoding();
```
missing `@Override`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,256 @@
+
+    /**
+     * Can only be called after the child class is initialized enough that getBlock will return the right value
+     * @return
```
remove the empty `@return`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractInterleavedBlock.java:
```diff
@@ -0,0 +1,256 @@
+            throw new IndexOutOfBoundsException("Invalid position (" + position + "), length (" + length + ") in block with " + positionCount + " positions");
+        }
+        if (position % columns != 0 || length % columns != 0) {
+            throw new IllegalArgumentException("position (" + position + ") and/or length (" + length + ") does not evenly divide columns (" + columns + ")");
```
See comment in tests about allowing this

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    /**
+     * Caller of this private constructor is responsible for making sure `values` and `offsets` are constructed with the same `blockBuilderStatus` as the one in the argument
```
comment out of date, as `offsets` is now a `Slice`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java:
```diff
@@ -0,0 +1,240 @@
+    private ArrayBlockBuilder(BlockBuilderStatus blockBuilderStatus, BlockBuilder values, SliceOutput offsets, SliceOutput valueIsNull)
+    {
+        this.blockBuilderStatus = requireNonNull(blockBuilderStatus, "blockBuilderStatus is null");
+        this.values = requireNonNull(values);
```
add second argument for the rest of these, just like the check for `blockBuilderStatus`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java:
```diff
@@ -0,0 +1,240 @@
+        }
+
+        entryAdded(false);
+        ongoingBlockEntry = false;
```
`ongoingBlockEntry` is never set to true, also I think you should remove `currentEntrySize` since it looks like it's superceded by `ongoingBlockEntry`

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockEncoding.java:
```diff
@@ -0,0 +1,98 @@
+        valueBlockEncoding.writeBlock(sliceOutput, arrayBlock.getValues());
+        int positionCount = arrayBlock.getPositionCount();
+        sliceOutput.appendInt(positionCount);
+        sliceOutput.appendBytes(arrayBlock.getOffsets().slice(0, (positionCount + 1) * 4));
```
is it necessary to have the `.slice()` call here? That looks like it should be a no-op

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayElementBlockWriter.java:
```diff
@@ -0,0 +1,159 @@
+
+    private void checkOngoingBlockEntry()
+    {
+        if (ongoingBlockEntry) {
```
this is never set to true, so I don't think this check is working. Also add a test to catch that

cberner replied:
I think you missed this comment

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockBuilder.java:
```diff
@@ -0,0 +1,284 @@
+    private int startSize;
+    private int retainedSizeInBytes;
+    private int startRetainedSize;
+    private boolean ongoingBlockEntry;
```
this is never set to true. add a test

cberner replied:
I think you missed this comment

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockBuilder.java:
```diff
@@ -0,0 +1,284 @@
+        this.startRetainedSize = -1;
+    }
+
+    private void incrementPositionCountAndCurrentBlockId()
```
this is only used in one place, so just inline it

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/InterleavedBlockBuilder.java:
```diff
@@ -0,0 +1,284 @@
+        return retainedSizeInBytes;
+    }
+
+    private void recordStartSizes(BlockBuilder blockBuilder)
```
name this `recordStartSizesIfNecessary`

cberner commented:
Back to you for some more comments. Mostly minor stuff, but you need tests that would have caught some of those

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java:
```diff
@@ -0,0 +1,237 @@
+    @Override
+    public Block getSingleValueBlock(int position)
+    {
+        checkReadablePosition(position);
```
implement this method as `return copyRegion(position, 1);`?

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java:
```diff
@@ -0,0 +1,247 @@
+        if (currentEntrySize != 0) {
+            throw new IllegalStateException("Expected entry size to be exactly " + 0 + " but was " + currentEntrySize);
+        }
+        currentEntrySize++;
```
I think you can remove `currentEntrySize` and set `ongoingBlockEntry = true` in this method

cberner commented:
Looks good, but make sure to address the comments I made
